headless/src/main.rs

```diff
@@ -1,58 +1,91 @@
-// Original functionality for profiling and debugging gridlock have been removed.
+lazy_static::lazy_static! {
+    static ref MAP: RwLock<Map> = RwLock::new(Map::blank());
+    static ref SIM: RwLock<Sim> = RwLock::new(Sim::new(&Map::blank(), SimOptions::new("tmp"), &mut Timer::throwaway()));
+}
 
-fn main() {
+#[tokio::main]
+async fn main() {
     let mut args = CmdArgs::new();
-    let num_days = args
-        .optional_parse("--days", |s| s.parse::<usize>())
-        .unwrap_or(1);
+    let mut sim_flags = SimFlags::from_args(&mut args);
+    let port = args.required("--port").parse::<u16>().unwrap();
     args.done();
 
-    let mut sim_flags = SimFlags::synthetic_test("montlake", "pandemic");
-    sim_flags.opts.enable_pandemic_model = Some(XorShiftRng::from_seed([sim_flags.rng_seed; 16]));
     // Less spam
     sim_flags.opts.alerts = AlertHandler::Silence;
     let mut timer = Timer::new("setup headless");
-    let (mut map, _, mut rng) = sim_flags.load(&mut timer);
-    map.hack_override_offstreet_spots(num_days);
-    // After the map changes, have to re-create the (empty) Sim, because things like
-    // ParkingSimState are out of sync.
-    let mut sim = Sim::new(&map, sim_flags.opts.clone(), &mut timer);
-
-    let base_scenario: Scenario = abstutil::read_binary(
-        abstutil::path_scenario(map.get_name(), "weekday"),
-        &mut timer,
-    );
-    ScenarioModifier::RepeatDays(num_days)
-        .apply(&map, base_scenario, &mut rng)
-        .instantiate(&mut sim, &map, &mut rng, &mut timer);
-    timer.done();
+    let (map, sim, _) = sim_flags.load(&mut timer);
+    *MAP.write().unwrap() = map;
+    *SIM.write().unwrap() = sim;
 
-    run_experiment(&map, &mut sim);
+    let addr = std::net::SocketAddr::from(([127, 0, 0, 1], port));
+    println!("Listening on http://{}", addr);
+    let serve_future = Server::bind(&addr).serve(hyper::service::make_service_fn(|_| async {
+        Ok::<_, hyper::Error>(hyper::service::service_fn(serve_req))
+    }));
+    if let Err(err) = serve_future.await {
+        panic!("Server error: {}", err);
+    }
 }
 
-fn run_experiment(map: &Map, sim: &mut Sim) {
-    let timer = Timer::new("run sim until done");
-    sim.run_until_done(
-        &map,
-        |sim, _map| {
-            // This'll run every 30 sim seconds
-            if false {
-                println!(
-                    "At {}, {} infected",
-                    sim.time(),
-                    prettyprint_usize(sim.get_pandemic_model().unwrap().count_infected())
-                );
-            }
-        },
-        None,
+async fn serve_req(req: Request<Body>) -> Result<Response<Body>, hyper::Error> {
+    let path = req.uri().path();
+    // Scrape query parameters
+    let params: HashMap<&str, &str> = req
+        .uri()
+        .query()
+        .unwrap_or("")
+        .split("&")
+        .filter(|x| !x.is_empty())
+        .map(|kv| {
+            let parts = kv.split("=").collect::<Vec<_>>();
```

Comment:
Alternatively, since we already rely on the URL crate, we could:

```
let uri_string = request.uri().to_string();
let request_url = Url::parse(&uri_string).unwrap();
let params = request_url.query_pairs();
```

But take it or leave it.

Reply:
Much nicer, done!